<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [3]:
!pip install ipython-sql

  Using cached typing_extensions-4.12.2-py3-none-any.whl.metadata (3.0 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 5.1 MB/s eta 0:00:00a 0:00:01
Using cached typing_extensions-4.12.2-py3-none-any.whl (37 kB)


In [4]:
import pandas as pd
import sqlite3
conn = sqlite3.connect('FinalDB.db')


Load the SQL magic module


In [5]:
%load_ext sql



Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [6]:
import pandas as pd
import sqlite3

# Establish SQLite connection
conn = sqlite3.connect('FinalDB.db')

# URLs for datasets
census_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv"
schools_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv"
crime_data_url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv"

# Load datasets into DataFrames
census_df = pd.read_csv(census_data_url)
schools_df = pd.read_csv(schools_data_url)
crime_df = pd.read_csv(crime_data_url)

# Save DataFrames to SQLite tables
census_df.to_sql("CENSUS_DATA", conn, if_exists='replace', index=False)
schools_df.to_sql("CHICAGO_PUBLIC_SCHOOLS", conn, if_exists='replace', index=False)
crime_df.to_sql("CHICAGO_CRIME_DATA", conn, if_exists='replace', index=False)


533

Establish a connection between SQL magic module and the database `FinalDB.db`


In [7]:
%sql sqlite:///FinalDB.db

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [9]:
# Using sqlite3 to execute the query and fetch the result
cursor = conn.cursor()
cursor.execute("SELECT COUNT(*) AS total_crimes FROM CHICAGO_CRIME_DATA;")
result = cursor.fetchone()

# Display the total number of crimes
print("Total number of crimes recorded:", result[0])



Total number of crimes recorded: 533


### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [10]:
# Query to find community areas with per capita income less than 11000
query = """
SELECT community_area_number, community_area_name
FROM CENSUS_DATA
WHERE per_capita_income < 11000;
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("Community Area Number:", row[0], "| Community Area Name:", row[1])


Community Area Number: 26.0 | Community Area Name: West Garfield Park
Community Area Number: 30.0 | Community Area Name: South Lawndale
Community Area Number: 37.0 | Community Area Name: Fuller Park
Community Area Number: 54.0 | Community Area Name: Riverdale


### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [11]:
# Query to find all case numbers for crimes involving minors
query = """
SELECT case_number
FROM CHICAGO_CRIME_DATA
WHERE description LIKE '%minor%' OR description LIKE '%juvenile%';
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("Case Number:", row[0])


Case Number: HL266884
Case Number: HK238408


### Problem 4

##### List all kidnapping crimes involving a child?


In [12]:
# Query to find all kidnapping crimes involving a child
query = """
SELECT case_number, primary_type, description
FROM CHICAGO_CRIME_DATA
WHERE primary_type = 'KIDNAPPING' AND description LIKE '%child%';
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("Case Number:", row[0], "| Primary Type:", row[1], "| Description:", row[2])


Case Number: HN144152 | Primary Type: KIDNAPPING | Description: CHILD ABDUCTION/STRANGER


### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


In [13]:
# Query to list distinct types of crimes recorded at schools
query = """
SELECT DISTINCT primary_type
FROM CHICAGO_CRIME_DATA
WHERE location_description LIKE '%school%';
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("Crime Type:", row[0])


Crime Type: BATTERY
Crime Type: CRIMINAL DAMAGE
Crime Type: NARCOTICS
Crime Type: ASSAULT
Crime Type: CRIMINAL TRESPASS
Crime Type: PUBLIC PEACE VIOLATION


### Problem 6

##### List the type of schools along with the average safety score for each type.


In [17]:
# Query to get column names from CHICAGO_PUBLIC_SCHOOLS table
query = "PRAGMA table_info(CHICAGO_PUBLIC_SCHOOLS);"

# Execute the query to see the table structure
cursor.execute(query)
columns = cursor.fetchall()

# Display column names
for column in columns:
    print("Column Name:", column[1])
    
# Updated query to find the type of schools along with their average safety score
query = """
SELECT "Elementary, Middle, or High School" AS school_type, AVG(SAFETY_SCORE) AS average_safety_score
FROM CHICAGO_PUBLIC_SCHOOLS
GROUP BY "Elementary, Middle, or High School";
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("School Type:", row[0], "| Average Safety Score:", row[1])
    



Column Name: School_ID
Column Name: NAME_OF_SCHOOL
Column Name: Elementary, Middle, or High School
Column Name: Street_Address
Column Name: City
Column Name: State
Column Name: ZIP_Code
Column Name: Phone_Number
Column Name: Link
Column Name: Network_Manager
Column Name: Collaborative_Name
Column Name: Adequate_Yearly_Progress_Made_
Column Name: Track_Schedule
Column Name: CPS_Performance_Policy_Status
Column Name: CPS_Performance_Policy_Level
Column Name: HEALTHY_SCHOOL_CERTIFIED
Column Name: Safety_Icon
Column Name: SAFETY_SCORE
Column Name: Family_Involvement_Icon
Column Name: Family_Involvement_Score
Column Name: Environment_Icon
Column Name: Environment_Score
Column Name: Instruction_Icon
Column Name: Instruction_Score
Column Name: Leaders_Icon
Column Name: Leaders_Score
Column Name: Teachers_Icon
Column Name: Teachers_Score
Column Name: Parent_Engagement_Icon
Column Name: Parent_Engagement_Score
Column Name: Parent_Environment_Icon
Column Name: Parent_Environment_Score
Column Nam

### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [18]:
# Query to list 5 community areas with the highest percentage of households below the poverty line
query = """
SELECT community_area_name, percent_households_below_poverty
FROM CENSUS_DATA
ORDER BY percent_households_below_poverty DESC
LIMIT 5;
"""

# Execute the query
cursor.execute(query)
results = cursor.fetchall()

# Display the results
for row in results:
    print("Community Area:", row[0], "| % Households Below Poverty:", row[1])


Community Area: Riverdale | % Households Below Poverty: 56.5
Community Area: Fuller Park | % Households Below Poverty: 51.2
Community Area: Englewood | % Households Below Poverty: 46.6
Community Area: North Lawndale | % Households Below Poverty: 43.1
Community Area: East Garfield Park | % Households Below Poverty: 42.4


### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [19]:
# Query to find the most crime-prone community area number
query = """
SELECT community_area_number
FROM CHICAGO_CRIME_DATA
GROUP BY community_area_number
ORDER BY COUNT(*) DESC
LIMIT 1;
"""

# Execute the query
cursor.execute(query)
result = cursor.fetchone()

# Display the most crime-prone community area number
print("Most Crime-Prone Community Area Number:", result[0])


Most Crime-Prone Community Area Number: 25.0



Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


In [20]:
# Query to find the name of the community area with the highest hardship index
query = """
SELECT community_area_name
FROM CENSUS_DATA
WHERE hardship_index = (SELECT MAX(hardship_index) FROM CENSUS_DATA);
"""

# Execute the query
cursor.execute(query)
result = cursor.fetchone()

# Display the community area with the highest hardship index
print("Community Area with Highest Hardship Index:", result[0])


Community Area with Highest Hardship Index: Riverdale


### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


In [21]:
# Query to find the community area name with the most number of crimes
query = """
SELECT community_area_name
FROM CENSUS_DATA
WHERE community_area_number = (
    SELECT community_area_number
    FROM CHICAGO_CRIME_DATA
    GROUP BY community_area_number
    ORDER BY COUNT(*) DESC
    LIMIT 1
);
"""

# Execute the query
cursor.execute(query)
result = cursor.fetchone()

# Display the community area with the most crimes
print("Community Area with Most Crimes:", result[0])


Community Area with Most Crimes: Austin


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
